### DISCUSSION TITLE
Creating models in a large project

### DISCUSSION THREAD

**Discussion by radqut** [CodinGame/monaco-vscode-api#477] [Q&A]
Hi everyone,

I have the large project (~100 ts files). I also have `typescript-language-server` running on the server.

I create a model of the `index.ts` file that contains:

```ts
import { print } from "./lib.ts";

print();
```

The monaco correct sees what is the `lib.ts` file, but when I press ctrl + click on the "./lib.ts", an error appears:

```ts
Error: Unable to read file '\workspace\lib.ts' (Error: Unable to resolve nonexistent file '\workspace\lib.ts')
```

instead of calling the registered editor opener:

```ts
monaco.editor.registerEditorOpener({
  openCodeEditor(source, resource, selectionOrPosition) {
    ...
  },
});
```

But when I add a model of the `lib.ts` file, the error doesn't appear, but I can't create models for all 100 files to get imports works.

Is there any way to get it works? I think there should be something lazy loading/creating models.

**Comment by CGNonofr** [comment]
Hello,

I will need more information on how you've integrated the lib

> I have the large project (~100 ts files). I also have typescript-language-server running on the server.

How do you communicate between the client and the server?

> I create a model of the index.ts file that contains:

How are you creating it? Are you creating the model directly or using a model reference?

Are you using the model service override?

> But when I add a model of the lib.ts file, the error doesn't appear, but I can't create models for all 100 files to get imports works.

Based on it, I have the feeling you don't use the model service override, are you creating the models using `monaco.editor.createModel`?

> Is there any way to get it works? I think there should be something lazy loading/creating models.

There is, all is expected is to have the model service override registered, and your files available on the virtual filesystem

**Reply by radqut** [reply]
Here is the basic project - [link](https://github.com/MichailShcherbakov/monaco-tls)

**Reply by CGNonofr** [reply]
It doesn't seem you interact at all with the remote filesystem.

The Language server probably needs the file to be up to date on its filesystem to operate properly.

What I recommend doing for simple scenarios:
- You extend the language server protocol and add methods to be able to write and read file remotely
- You implement a custom file system provider, relying on those methods
- You make sure the files are up to date on the server
- You can now use `createModelReference` to open the remote file, and it will also be saved when changed

If you need your app to work even when your server is down:
- You extend the language server protocol and add methods to be able to write and read file remotely
- You register all your files client-side in a RegisteredFileSystemProvider and register it
- You listen on file changes client-side (can be done using the VSCode api) and update the files on the server when they change
- (Optional) You would probably still need to have a custom file system provider able to fetch files on the server, for example for library files (when hitting ctrl+click on a library)

**Reply by radqut** [reply]
Can I just to use `RegisteredUriFile` to fetch the external files by  "http://my-server/files/index.ts" url? I just want the files to be loaded from my server and so that no the error occurs:

```
Error: Unable to read file '\workspace\lib.ts' (Error: Unable to resolve nonexistent file '\workspace\lib.ts')
```

**Reply by CGNonofr** [reply]
Sure, but you would need to register all files one by one

**Reply by radqut** [reply]
It seems to be a vicious circle. So there is no way to use 'Go to definition' and '@codingame/monaco-vs code-typescript-language-features-default-extension' without force including the file whose path is specified in the import statement?

**Reply by CGNonofr** [reply]
What do you mean by `force including`

What is the issue with implementing your own file system provider?

**Reply by radqut** [reply]
> What do you mean by force including

Creating models for all files (10, 100 or 1000)

> What is the issue with implementing your own file system provider?

there are no examples or documentation

**Reply by CGNonofr** [reply]
> there are no examples or documentation

Probably :thinking: 

What I suggest is instead of using `RegisteredFileSystemProvider`, implement your own `IFileSystemProviderWithFileReadWriteCapability`:

```typescript
import * as vscode from 'vscode'
import { IFileSystemProviderWithFileReadWriteCapability, FileSystemProviderCapabilities } from '@codingame/monaco-vscode-files-service-override'

class MyCustomProvider implements IFileSystemProviderWithFileReadWriteCapability {
  capabilities = FileSystemProviderCapabilities.FileReadWrite | FileSystemProviderCapabilities.PathCaseSensitive
  onDidChangeCapabilities = new vscode.EventEmitter<void>().event
  onDidChangeFile = new vscode.EventEmitter<readonly IFileChange[]>().event
  readFile(resource: URI): Promise<Uint8Array> {
    throw new Error('Method not implemented.')
  }
  writeFile(resource: URI, content: Uint8Array, opts: IFileWriteOptions): Promise<void> {
    throw new Error('Method not implemented.')
  }
  watch(resource: URI, opts: IWatchOptions): IDisposable {
    throw new Error('Method not implemented.')
  }
  stat(resource: URI): Promise<IStat> {
    throw new Error('Method not implemented.')
  }
  mkdir(resource: URI): Promise<void> {
    throw new Error('Method not implemented.')
  }
  readdir(resource: URI): Promise<[string, FileType][]> {
    throw new Error('Method not implemented.')
  }
  delete(resource: URI, opts: IFileDeleteOptions): Promise<void> {
    throw new Error('Method not implemented.')
  }
  rename(from: URI, to: URI, opts: IFileOverwriteOptions): Promise<void> {
    throw new Error('Method not implemented.')
  }
}
```
